File: code/WRSN-CEC2021-MMC-MFEA/src/elements/Charger.java
package elements;

import java.util.ArrayList;
import java.util.List;

import problem.Configs;
import problem.ProblemManager;

public class Charger {
	private double E0;
	private double speed;
	private double Pm;
	private double U;

	public Charger(double e0, double speed, double pm, double u) {
		this.setE0(e0);
		this.setPm(pm);
		this.setSpeed(speed);
		this.setU(u);
	}

	public double getMovingTime(List<Integer> path) {
		double res = 0;
		res += ProblemManager.distance[ProblemManager.serviceStation.getId()][path.get(0)];
		res += ProblemManager.distance[path.get(path.size() - 1)][ProblemManager.serviceStation.getId()];
		for (int i = 1; i < path.size(); i++) {
			res += ProblemManager.distance[path.get(i - 1)][path.get(i)];
		}
		return res / this.speed;
	}

	public double getMaxChargingTime(double travelingDistance) {
		return (this.getE0() - travelingDistance * this.getPm() / getSpeed()) / this.getU();
	}

	public double getMaxChargingTime(ArrayList<Integer> path) {
		double movingTime = this.getMovingTime(path);
		if (movingTime > Configs.T) {
			System.err.println("Travling time exceed charging cycle! " +  movingTime);
			System.exit(0);
		}
		double ct = (this.getE0() - movingTime * this.getPm()) / this.getU();
		return Math.min(ct, Configs.T - movingTime);

	}

	public double getMaxChargingTime(int[] path) {
		double movingTime = this.getMovingTime(path);
		if (movingTime > Configs.T) {
			System.err.println("Travling time exceed charging cycle!");
			System.exit(0);
		}
		double ct = (this.getE0() - movingTime * this.getPm()) / this.getU();
		return Math.min(ct, Configs.T - movingTime);

	}

	public double getMovingTime(int[] path) {
		double res = 0;
		int n = path.length;
		res += ProblemManager.distance[ProblemManager.serviceStation.getId()][path[0]];
		res += ProblemManager.distance[path[n - 1]][ProblemManager.serviceStation.getId()];
		for (int i = 1; i < n; i++) {
			res += ProblemManager.distance[path[i - 1]][path[i]];
		}
		return res / this.speed;
	}

	public double getEmove(int[] path) {
		return this.getMovingTime(path) * this.Pm;
	}

	public double getE0() {
		return E0;
	}

	public void setE0(double e0) {
		E0 = e0;
	}

	public double getSpeed() {
		return speed;
	}

	public void setSpeed(double speed) {
		this.speed = speed;
	}

	public double getPm() {
		return Pm;
	}

	public void setPm(double pm) {
		Pm = pm;
	}

	public double getU() {
		return U;
	}

	public void setU(double u) {
		U = u;
	}

}

File: elements/Charger.py
import sys
from multipledispatch import dispatch
try:
    from problem.ProblemManager import ProblemManager
except ImportError:
    import sys
    ProblemManager = sys.modules['problem' + '.ProblemManager']
from problem.Configs import Configs

class Charger:
    
    E0: float
    """Charger's energy capacity"""
    
    speed: float
    """Charger's travel speed"""

    Pm: float
    """Charger's per-second traveling consumption rate"""

    U: float
    """Charger's charging rate"""

    def __init__(self, e0: float, speed: float, pm: float, u: float) -> None:
        self.E0 = e0
        self.speed = speed
        self.Pm = pm
        self.U = u

    def getMovingTime(self, path:'list[int]') -> float:
        """Calculate charger's moving time given a path (list of node to travel to)"""
        pathLength = 0.0
        pathLength += ProblemManager.distance[ProblemManager.serviceStation.getId()][path[0]]
        pathLength += ProblemManager.distance[path[len(path) - 1]][ProblemManager.serviceStation.id]
        for i in range(1,len(path)):
            pathLength += ProblemManager.distance[path[i-1]][path[i]]
        return pathLength/self.speed

    @dispatch(float)
    def getMaxChargingTime(self, travelingDistance: float) -> float: # type: ignore
        """Calculate maximum time that this charger can charge given total moving distance"""
        return (self.E0 - travelingDistance * self.Pm / self.speed) / self.U

    @dispatch(list)
    def getMaxChargingTime(self, path):  # type: ignore
        """Calculate maximum time that this charger can charge in a given moving path\n
            Can not exceed maximum time in config"""
        movingTime = self.getMovingTime(path)
        if movingTime > Configs.T:
            print(f"Traveling time exceed charging cycle! {movingTime}", file=sys.stderr)
            sys.exit(0)

        #( Availale energy - moving time * moving consumption ) / charge rate
        ct = (self.E0 - movingTime*self.Pm) / self.U
        return min(ct, Configs.T - movingTime)

    def getEmove(self, path):
        """Calculate total energy used for moving in a given moving path"""
        return self.getMovingTime(path) * self.Pm

    def getU(self):
        return self.U

